File: app/src/main/java/com/battlelancer/seriesguide/util/tasks/BaseShowActionTask.java
package com.battlelancer.seriesguide.util.tasks;

import android.content.Context;
import androidx.annotation.NonNull;
import com.battlelancer.seriesguide.SgApp;
import com.battlelancer.seriesguide.traktapi.TraktCredentials;
import com.uwetrottmann.trakt5.entities.ShowIds;
import com.uwetrottmann.trakt5.entities.SyncItems;
import com.uwetrottmann.trakt5.entities.SyncResponse;
import com.uwetrottmann.trakt5.entities.SyncShow;
import com.uwetrottmann.trakt5.services.Sync;
import org.greenrobot.eventbus.EventBus;
import retrofit2.Call;

public abstract class BaseShowActionTask extends BaseActionTask {

    public static class ShowChangedEvent {
    }

    private final int showTmdbId;

    public BaseShowActionTask(Context context, int showTmdbId) {
        super(context);
        this.showTmdbId = showTmdbId;
    }

    @Override
    protected boolean isSendingToHexagon() {
        return false;
    }

    @Override
    protected Integer doBackgroundAction(Void... params) {
        if (isSendingToTrakt()) {
            if (!TraktCredentials.get(getContext()).hasCredentials()) {
                return ERROR_TRAKT_AUTH;
            }

            SyncItems items = new SyncItems().shows(new SyncShow().id(ShowIds.tmdb(showTmdbId)));
            Sync traktSync = SgApp.getServicesComponent(getContext()).traktSync();

            int result = executeTraktCall(buildTraktCall(traktSync, items), getTraktAction(),
                    body -> {
                        if (isShowNotFound(body)) {
                            return ERROR_TRAKT_API_NOT_FOUND;
                        } else {
                            return SUCCESS;
                        }
                    });
            //noinspection RedundantIfStatement
            if (result != SUCCESS) {
                return result;
            }
        }

        return SUCCESS;
    }

    private static boolean isShowNotFound(SyncResponse response) {
        // if show was not found on trakt
        return response.not_found != null && response.not_found.shows != null
                && response.not_found.shows.size() != 0;
    }

    @Override
    protected void onPostExecute(Integer result) {
        super.onPostExecute(result);

        if (result == SUCCESS) {
            EventBus.getDefault().post(new ShowChangedEvent());
        }
    }

    @NonNull
    protected abstract String getTraktAction();

    @NonNull
    protected abstract Call<SyncResponse> buildTraktCall(Sync traktSync, SyncItems items);
}

File: app/src/main/java/com/battlelancer/seriesguide/util/tasks/BaseShowActionTask.kt
// SPDX-License-Identifier: GPL-3.0-or-later
// SPDX-FileCopyrightText: Copyright © 2016 Uwe Trottmann <uwe@uwetrottmann.com>

package com.battlelancer.seriesguide.util.tasks

import android.content.Context
import com.battlelancer.seriesguide.SgApp
import com.battlelancer.seriesguide.traktapi.TraktCredentials
import com.uwetrottmann.trakt5.entities.ShowIds
import com.uwetrottmann.trakt5.entities.SyncItems
import com.uwetrottmann.trakt5.entities.SyncResponse
import com.uwetrottmann.trakt5.entities.SyncShow
import com.uwetrottmann.trakt5.services.Sync
import org.greenrobot.eventbus.EventBus
import retrofit2.Call

abstract class BaseShowActionTask(
    context: Context,
    private val showTmdbId: Int
) : BaseActionTask(context) {

    class ShowChangedEvent

    override val isSendingToHexagon: Boolean
        get() = false

    override suspend fun doBackgroundAction(): Int {
        if (isSendingToTrakt) {
            if (!TraktCredentials.get(context).hasCredentials()) {
                return ERROR_TRAKT_AUTH
            }

            val items = SyncItems().shows(SyncShow().id(ShowIds.tmdb(showTmdbId)))
            val trakt = SgApp.getServicesComponent(context).trakt()
            val traktSync = trakt.sync()

            val result = executeTraktCall(
                buildTraktCall(traktSync, items),
                trakt,
                traktAction,
                object : ResponseCallback<SyncResponse> {
                    override fun handleSuccessfulResponse(body: SyncResponse): Int {
                        return if (isShowNotFound(body)) {
                            ERROR_TRAKT_API_NOT_FOUND
                        } else {
                            SUCCESS
                        }
                    }
                })
            if (result != SUCCESS) {
                return result
            }
        }

        return SUCCESS
    }

    private fun isShowNotFound(response: SyncResponse): Boolean {
        // if show was not found on trakt
        return response.not_found?.shows?.isNotEmpty()
                ?: false
    }

    override fun onPostExecute(result: Int) {
        super.onPostExecute(result)

        if (result == SUCCESS) {
            EventBus.getDefault().post(ShowChangedEvent())
        }
    }

    protected abstract val traktAction: String

    protected abstract fun buildTraktCall(traktSync: Sync, items: SyncItems): Call<SyncResponse>

}
